presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionOperator.java

```diff
@@ -0,0 +1,250 @@
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.sql.planner.PlanFragment;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class NativeExecutionOperator
+        implements SourceOperator, WrapperSourceOperator
+{
+    private final PlanNodeId sourceId;
+    private final OperatorContext operatorContext;
+    private final LocalMemoryContext systemMemoryContext;
+    private final JsonCodec<TaskSource> taskSourceCodec;
+    private final JsonCodec<PlanFragment> planFragmentCodec;
+    private final JsonCodec<TableWriteInfo> tableWriteInfoCodec;
+    private final PlanFragment planFragment;
+    private final TableWriteInfo tableWriteInfo;
+    private TaskSource taskSource;
+    private final boolean isFirstOperator;
+
+    private boolean finished;
+
+    public NativeExecutionOperator(
+            PlanNodeId sourceId,
+            OperatorContext operatorContext,
+            JsonCodec<TaskSource> taskSourceCodec,
+            JsonCodec<PlanFragment> planFragmentCodec,
+            JsonCodec<TableWriteInfo> tableWriteInfoCodec,
+            PlanFragment planFragment,
+            TableWriteInfo tableWriteInfo,
+            boolean isFirstOperator)
+    {
+        this.sourceId = requireNonNull(sourceId, "sourceId is null");
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.systemMemoryContext = operatorContext.localSystemMemoryContext();
+        this.taskSourceCodec = requireNonNull(taskSourceCodec, "taskSourceJsonCodec is null");
+        this.planFragmentCodec = requireNonNull(planFragmentCodec, "planFragmentCodec is null");
+        this.tableWriteInfoCodec = requireNonNull(tableWriteInfoCodec, "tableWriteInfoCodec is null");
+        this.planFragment = requireNonNull(planFragment, "planFragment is null");
+        this.tableWriteInfo = requireNonNull(tableWriteInfo, "tableWriteInfo is null");
+        this.taskSource = null;
+        this.finished = false;
+        this.isFirstOperator = isFirstOperator;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return false;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (finished) {
+            return null;
+        }
+        finished = true;
+        return null;
```

Comment:
I would add TODOs here to explain what logic do we expect to add in the future. Same for other methods.

I guess we will have page serde happening just like exchangeOperator